### PR TITLE
Refactor dadadj for CCPP framework 

### PR DESCRIPTION
**Pull request by jtruesdal**
Add dadadj ccpp routine meta data and suite xml files to atmos_phys.

### CODE REVIEW

File: suite_dadadj.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (nusbaume) [inline suite_dadadj.xml:1]: Given that `dadadj` is the first true CAM7 physics scheme, I might vote to make a copy of this SDF called `suite_cam7.xml`.  I then might move the `suite_dadadj.xml` file to a new `test/test_sdfs` directory, as we will likely only use this particular SDF for testing purposes.

**Review by nusbaume — CHANGES_REQUESTED** [review]
Great work @jtruesdal !  I do have some change requests, but  please let me know if you have any questions or concerns with any of them.  Thanks!

File: dadadj/dadadj.F90 [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    logical :: dodad(ncol)     ! .TRUE. ==> do dry adjustment
+
+    !-----------------------------------------------------------------------
+
+    zeps = 2.0e-5_kind_phys           ! set convergence criteria
+    errmsg = ''
+    errflg = 0
+    scheme_name = 'DADADJ'
+
+    allocate(c1dad(nlvdry), c2dad(nlvdry), c3dad(nlvdry), c4dad(nlvdry))
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.F90]: Added all the stat checks and endruns.

File: dadadj/dadadj.F90 [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    t = state_t
+    q = state_q
+
+    ! Find gridpoints with unstable stratification
+
+    do i = 1, ncol
+       cappaint = 0.5_kind_phys*(cappa(i,2) + cappa(i,1))
+       gammad = cappaint*0.5_kind_phys*(t(i,2) + t(i,1))/pint(i,2)
+       dtdp = (t(i,2) - t(i,1))/(pmid(i,2) - pmid(i,1))
+       dodad(i) = (dtdp + zeps) .gt. gammad
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.F90]: Done.

File: dadadj/dadadj.F90 [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+               end do
+
+               if (ilconv) cycle COL ! convergence => next longitude
+
+            end do
+
+            zeps = zeps + zeps
+            if (zeps > 1.e-4_kind_phys) then
+               errflg = i
+               return                ! error return
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.F90]: Yes. Done.

File: dadadj/dadadj.meta [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+  standard_name = ccpp_error_message
+  long_name = Error message for error handling in CCPP
+  units = none
+  type = character | kind = len=512
+  dimensions = ()
+  intent = out
+[ errflg ]
+  standard_name = ccpp_error_flag
+  long_name = Error flag for error handling in CCPP
+  units = flag
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Cone.

[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Done

File: dadadj/dadadj.meta [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  dimensions = (horizontal_loop_extent,vertical_layer_dimension)
+  intent = in
+[ state_t ]
+  standard_name = air_temperature
+  type = real | kind = kind_phys
+  units = K
+  dimensions = (horizontal_dimension, vertical_layer_dimension)
+  intent = in
+[ state_q ]
+  standard_name = water_vapor_mixing_ratio_wrt_dry_air
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Done.

File: dadadj/dadadj.meta [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  intent = out
+[ q_tend ]
+  standard_name = tendency_of_water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
+  units = kg kg-1 s-1
+  type = real | kind = kind_phys
+  dimensions = (horizontal_dimension, vertical_layer_dimension)
+  intent = out
+[ dadpdf ]
+  standard_name = indicator_for_dry_adiabatic_adjusted_grid_cell
+  units = 1
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Done

File: dadadj/dadadj.meta [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+  standard_name = air_pressure_thickness
+  units = Pa
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent,vertical_layer_dimension)
+  intent = in
+[ state_t ]
+  standard_name = air_temperature
+  type = real | kind = kind_phys
+  units = K
+  dimensions = (horizontal_dimension, vertical_layer_dimension)
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Renamed all the instances of horizontal_dimension to horizontal_loop_extent

File: dadadj/dadadj.meta [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  standard_name = ccpp_error_message
+  long_name = Error message for error handling in CCPP
+  units = none
+  type = character | kind = len=512
+  dimensions = ()
+  intent = out
+[ errflg ]
+  standard_name = ccpp_error_flag
+  long_name = Error flag for error handling in CCPP
+  units = flag
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Done.

File: dadadj/dadadj_namelist.xml [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+<!-- DADADJ configuration options -->
+
+<entry id="dadadj_nlvdry">
+  <type>integer</type>
+  <category>dry_conv_adj</category>
+  <group>dadadj_nl</group>
+  <standard_name>number_of_vertical_levels_from_model_top_where_dry_adiabatic_adjustment_occurs</standard_name>
+  <units>count</units>
+  <desc>
+    Number of layers from the top of the model over which to do dry convective adjustment.
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj_namelist.xml]: Done.

File: dadadj/dadadj_namelist.xml [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+</entry>
+<entry id="dadadj_niter">
+  <type>integer</type>
+  <category>dry_conv_adj</category>
+  <group>dadadj_nl</group>
+  <standard_name>number_of_iterations_for_dry_adiabatic_adjustment_algorithm_convergence</standard_name>
+  <units>count</units>
+  <desc>
+    The maximum number of iterations to achieve convergence in dry adiabatic adjustment.
+    For WACCM-X it can be advantageous to use a number which is much higher than the CAM
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj_namelist.xml]: Done

File: doc/ChangeLog [hunk 12]
```diff
@@ -1,6 +1,52 @@
 
 ===============================================================
 
+Tag name: atmos_phys0_02_xxx
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:4]: Done.

File: doc/ChangeLog [hunk 13]
```diff
@@ -1,6 +1,52 @@
...
+	--- /project/amp/jet/collections/cam6_3_160_dryadj.042924.1856/src/dynamics/se/dyn_comp.F90	2024-05-02 16:56:49.746341064 -0600
+	+++ SourceMods/src.cam/dyn_comp.F90	2024-05-02 16:37:13.171433366 -0600
+	@@ -1429,6 +1429,7 @@
+          do j = 1, np
+             do i = 1, np
+                elem(ie)%state%T(i,j,:,1) = dbuf3(indx,:,ie)
++               elem(ie)%state%T(i,j,2,1) = dbuf3(indx,2,ie) + 60._r8
+                indx = indx + 1
+             end do
+          end do
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:42]: Created a dry_adiabatic_adjust directory and put the modified file underneath.

File: suite_dadadj.xml (hunk 1 above)
Comment (jtruesdal) [reply suite_dadadj.xml:1]: Temporarily left a suite_dadadj.xml in the original spot for testing but will remove before commit.  Updated the rest as recommended.

File: dry_adiabatic_adjust/dadadj.F90 [hunk 14]
```diff
@@ -38,7 +38,8 @@ subroutine dadadj_init(dadadj_nlvdry, dadadj_niter, nz, errmsg, errflg)
 
     if (dadadj_nlvdry >= nz .or. dadadj_nlvdry < 0) then
        errflg = 1
-       write(errmsg,*) 'dadadj_init: dadadj_nlvdry=',dadadj_nlvdry,' but must be less than the number of vertical levels '
+       write(errmsg,*) 'dadadj_init: dadadj_nlvdry=',dadadj_nlvdry,' but must be less than the number of vertical levels ',&
+       '(',nz,'), and must be a positive integer.`
```
Comment (nusbaume) [inline dry_adiabatic_adjust/dadadj.F90:42]: Whoops, I think I had a typo for this particular suggestion:

```suggestion
       '(',nz,'), and must be a positive integer.'
```

File: suite_cam7.xml [hunk 15]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<suite name="dadadj" version="1.0">
+  <group name="physics_before_coupler">
+    <scheme>dadadj</scheme>
```
Comment (nusbaume) [inline suite_cam7.xml:5]: I might add a comment here stating that this is dry adiabatic adjustment:

```suggestion
    <!-- Dry Adiabatic Adjustment -->
    <scheme>dadadj</scheme>
```

File: dry_adiabatic_adjust/dadadj.meta [hunk 16]
```diff
@@ -126,9 +126,9 @@
   dimensions = ()
   intent = out
 [ errflg ]
-  standard_name = ccpp_error_flag
-  long_name = Error flag for error handling in CCPP
-  units = flag
+  standard_name = ccpp_error_code
+  long_name = Error code for error handling in CCPP
+  units = 1
```
Comment (nusbaume) [inline dry_adiabatic_adjust/dadadj.meta:131]: Just adding a second reminder to add an updated version of the `NamesNotInDictionary.txt` file to this PR following the instructions found online here:

https://github.com/ESCOMP/atmospheric_physics/wiki/Development-workflow#namesnotindictionarytxt

**Review by nusbaume — CHANGES_REQUESTED** [review]
Super-close!  I just have a few last (hopefully very minor) requests.

File: suite_cam7.xml [hunk 17]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<suite name="dadadj" version="1.0">
```
Comment (nusbaume) [inline suite_cam7.xml:3]: Whoops, forgot to mention in my previous review that this should probably be changed to `cam7` as well:

```suggestion
<suite name="cam7" version="1.0">
```

File: dadadj/dadadj.meta [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+  type = integer
+  dimensions = ()
+  intent = out
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply dadadj/dadadj.meta]: Thanks updated.

File: dry_adiabatic_adjust/dadadj.F90 (hunk 14 above)
Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj.F90]: Done.

File: suite_cam7.xml (hunk 15 above)
Comment (jtruesdal) [reply suite_cam7.xml:6]: Done.

File: dry_adiabatic_adjust/dadadj.meta (hunk 16 above)
Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj.meta:131]: Done.

**Review by nusbaume — APPROVED** [review]
Everything looks great now, thanks!

File: dry_adiabatic_adjust/dadadj.F90 [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+   ! Note: nlvdry ****MUST**** be < pver
+
+   COL: do i = 1, ncol
+
+      if (dodad(i)) then
+
+         zeps = 2.0e-5_kind_phys
+
+         do k = 1, nlvdry
+            cappaint = 0.5_kind_phys*(cappa(i,k+1) + cappa(i,k))
```
Comment (cacraigucar) [inline dry_adiabatic_adjust/dadadj.F90:166]: Was the addition of this line a bug that needed to be fixed? Does it not change answers between this and what is currently in CAM?

**Review by cacraigucar — CHANGES_REQUESTED** [review]

File: doc/ChangeLog [hunk 20]
```diff
@@ -1,6 +1,52 @@
...
+Code reviewed by:
+
+List all existing files that have been added (A), modified (M), or deleted (D),
+and describe the changes:
+
+A	dadadj/dadadj.F90  - minor refactor to the cam routine - ccpp'ize
+A	dadadj/dadadj.meta
+A	dadadj/dadadj_namelist.xml
+M	doc/ChangeLog
+A	suite_dadadj.xml - CCPP suite file
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:23]: changed the directory names to dry_adiabatic_adjust.  Is that what you meant?

File: doc/ChangeLog [hunk 21]
```diff
@@ -1,6 +1,52 @@
...
+    - dadadj.F90 routine was slightly refactored to update the logic and syntax
+
+Code reviewed by:
+
+List all existing files that have been added (A), modified (M), or deleted (D),
+and describe the changes:
+
+A	dadadj/dadadj.F90  - minor refactor to the cam routine - ccpp'ize
+A	dadadj/dadadj.meta
+A	dadadj/dadadj_namelist.xml
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:21]: Changed the directory names to dry_adiabatic_adjust

File: doc/ChangeLog [hunk 22]
```diff
@@ -1,6 +1,52 @@
...
+M	doc/ChangeLog
+A	suite_dadadj.xml - CCPP suite file
+
+
+List and Describe any test failures:
+    - Tested in CAM
+	Because dadadj doesn't normally get tripped I modified the Temp
+	profile to create an instability where the code would be exercized.
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:29]: corrected

File: doc/ChangeLog [hunk 23]
```diff
@@ -1,6 +1,52 @@
...
+	The snapshot files continain this instability.  The modification adds
```
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:30]: corrected

File: dry_adiabatic_adjust/dadadj.F90 (hunk 19 above)
Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj.F90:166]: I think it is a bug, introduced when variable cappa was added to the model.  I've sent this off to Peter to see if he agrees and if he needs to see a control before this is approved.

Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj.F90:166]: Peter approved the change and doesn't think we need a control.

File: doc/ChangeLog (hunk 20 above)
[the remark this answers is not in the document]
Comment (cacraigucar) [reply doc/ChangeLog:23]: No - the suite name changed and needs to be changed here as well

**Review by cacraigucar — CHANGES_REQUESTED** [review]

File: doc/ChangeLog (hunk 20 above)
[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:23]: Done.

**Comment by jtruesdal** [comment]
I've added a new temporary update routine to update the q tendency. 
 This will be removed once a standard utility for updating the constituents is in place.  I also modified the parameterization run method to return a heating tendency which now takes care of a roundoff issue in temperature.  Everything is again bfb.

**Comment by jtruesdal** [comment]
@cacraigucar @nusbaume I added a new temporary run method to update the constituent array (dadadj_apply_qv_tendency.F90/meta).  Everything is working so hopefully this is the final review.  I'm placing these snapshot files in their proper location.

File: doc/ChangeLog (hunk 20 above)
[the remark this answers is not in the document]
Comment (cacraigucar) [reply doc/ChangeLog:23]: Do you need to push the ChangeLog back?

[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:23]: I think this is OK.  When I first looked at it via this page it looked old but I refreshed and clicked on the file link via the lefthand frame and the names look to be spelled out.  I'm not sure why it wasn't updated the first time.

[the remark this answers is not in the document]
Comment (cacraigucar) [reply doc/ChangeLog:23]: The line here says the file is suite_dryadj.xml.  The file that I see was added is suite_dry_adiabatic_adjust.xml.  I looked directly in your doc/ChangeLog to see what it says.  I think the filename listed in the ChangeLog needs to be updated.

[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog:23]: Also the suite_dry_adiabatic_adjust.xml routine was moved to test/test_sdfs as requested by @nusbaume.  The suite_cam7.xml now contains dry_adiabatic_adjust as the first parameterization of the cam7 suite.  I think all future cam7 parameterizations will go into that suite as well as they are finished.  I'm happy to move things aound though if the plan has changed.

[the remark this answers is not in the document]
Comment (cacraigucar) [reply doc/ChangeLog:23]: I agree.  This is just a reminder that this section of the ChangeLog needs to be updated to reflect the current location and names of the files.

[the remark this answers is not in the document]
Comment (jtruesdal) [reply doc/ChangeLog]: Thanks for being patient. Updated Changelog with description and location of all 9 files, merged up to latest tag and regenerated NamesNotInDictionary.txt.

File: doc/ChangeLog [hunk 24]
```diff
@@ -1,5 +1,55 @@
 ===============================================================
 
+Tag name: atmos_phys0_0x_xxx
```
Comment (nusbaume) [inline doc/ChangeLog:3]: Just FYI that the tag will likely be `atmos_phys0_04_000`.

File: doc/ChangeLog [hunk 25]
```diff
@@ -1,5 +1,55 @@
...
+Originator(s): jet
+Date: May 3, 2024
+One-line Summary: dadadj CCPP mods
+Github PR URL: hhttps://github.com/ESCOMP/atmospheric_physics/pull/
```
Comment (nusbaume) [inline doc/ChangeLog:7]: Just a reminder to update this URL:

```suggestion
Github PR URL: https://github.com/ESCOMP/atmospheric_physics/pull/91
```

File: dry_adiabatic_adjust/dadadj.F90 [hunk 26]
```diff
@@ -50,22 +50,24 @@ end subroutine dadadj_init
...
     real(kind_phys), intent(in) :: pmid(:,:)   ! pressure at model levels
     real(kind_phys), intent(in) :: pint(:,:)   ! pressure at model interfaces
     real(kind_phys), intent(in) :: pdel(:,:)   ! vertical delta-p
     real(kind_phys), intent(in) :: cappa(:,:)  ! variable Kappa
+    real(kind_phys), intent(in) :: cpair(:,:)  ! heat capacity of air
     real(kind_phys), intent(in) :: state_t(:,:)   ! temperature (K)
     real(kind_phys), intent(in) :: state_q(:,:)   ! specific humidity
-    real(kind_phys), intent(out), target :: t_tend(:,:)   ! temperature tendency
-    real(kind_phys), intent(out), target :: q_tend(:,:)   ! specific humidity tendency
+    real(kind_phys), intent(out) :: s_tend(:,:)   ! temperature tendency
```
Comment (nusbaume) [inline dry_adiabatic_adjust/dadadj.F90:69]: I would change this comment to state that the tendency is for dry air enthalpy:

```suggestion
    real(kind_phys), intent(out) :: s_tend(:,:)   ! dry air enthalpy tendency
```

**Review by cacraigucar — APPROVED** [review]

File: dry_adiabatic_adjust/dadadj_apply_qv_tendency.meta [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+  type = scheme
+[ q_tend ]
+  standard_name = tendency_of_water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
+  units = kg kg-1 s-1
+  type = real | kind = kind_phys
+  dimensions = (horizontal_loop_extent, vertical_layer_dimension)
+  intent = in
+[ state_q ]
+  standard_name = water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
+  long_name = mass mixing ratio of water vapor / dry air
```
Comment (nusbaume) [inline dry_adiabatic_adjust/dadadj_apply_qv_tendency.meta:16]: I think this long name needs to be updated:

```suggestion
  long_name = mass mixing ratio of water vapor / moist air and condensed water
```

**Review by nusbaume — APPROVED** [review]
Looks good!  I just some minor requests to update a comment, a long name, and the ChangeLog (doesn't need a re-review though).

File: doc/ChangeLog (hunk 24 above)
Comment (jtruesdal) [reply doc/ChangeLog:3]: Thx. Updated.

File: doc/ChangeLog (hunk 25 above)
Comment (jtruesdal) [reply doc/ChangeLog]: Updated and added reviewers.

File: dry_adiabatic_adjust/dadadj.F90 (hunk 26 above)
Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj.F90]: Fixed comment.

File: dry_adiabatic_adjust/dadadj_apply_qv_tendency.meta (hunk 27 above)
Comment (jtruesdal) [reply dry_adiabatic_adjust/dadadj_apply_qv_tendency.meta]: Updated long_name

**Pull request merged**